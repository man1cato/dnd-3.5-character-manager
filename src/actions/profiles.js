import database from '../firebase/firebase';


//SET PROFILE IDS IN STORE
export const setProfiles = (profiles) => ({
    type: 'SET_PROFILES',
    profiles
})

//READ PROFILES FROM FIREBASE 
export const startGetProfiles = (uid) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles`).once('value').then((snapshot) => {
            let profiles = snapshot.val()
            if (profiles) {
                const profileIds = Object.keys(profiles)
                profiles = profileIds.map((id) => ({
                    id,
                    name: profiles[id].name,
                    race: profiles[id].race,
                    gender: profiles[id].gender,
                    jobClass: profiles[id].jobClass,
                    alignment: profiles[id].alignment,
                    level: profiles[id].level,
                    iconUrl: profiles[id].iconUrl
                }))
                dispatch(setProfiles(profiles))
            }
        })
    }
}

//REMOVE PROFILES FROM STORE
export const removeProfiles = () => ({
    type: 'REMOVE_PROFILES'
})