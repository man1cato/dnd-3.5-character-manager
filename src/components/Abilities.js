import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
