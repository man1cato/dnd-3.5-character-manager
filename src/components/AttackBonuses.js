import React, { useState, useEffect, Fragment } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { calcBaseMeleeBonus, calcBaseRangedBonus, calcBaseGrappleBonus, convertInputValue } from '../utils/utils'


const AttackBonuses = ({ profileAttackBonusMods, abilities, baseAttackBonuses, size, handleUpdate }) => {  
    const attackBonusBases = {
        melee: _.map(baseAttackBonuses, (bab) => calcBaseMeleeBonus(bab, abilities.str.score, size)),
        ranged: _.map(baseAttackBonuses, (bab) => calcBaseRangedBonus(bab, abilities.dex.score, size)),
        grapple: _.map(baseAttackBonuses, (bab) => calcBaseGrappleBonus(bab, abilities.str.score, size))
    }
    const [attackBonusMods, setAttackBonusMods] = useState(_.mapValues(profileAttackBonusMods, (val) => val || '')) 
    const [attackBonusTotals, setAttackBonusTotals] = useState(
        _.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val ))
    ) 
    
    useEffect(() => {
        setAttackBonusTotals(_.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val)))
        handleUpdate({ attackBonusMods })
    }, [attackBonusMods])

    return (    
        <div className="grid--6col">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(['melee', 'ranged', 'grapple'], (key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{attackBonusBases[key].join(' / ')}</div>
                    <div>+</div>
                    <input
                        type="number"
                        id={key}
                        value={attackBonusMods[key]}
                        onChange={(e) => {
                            setAttackBonusMods(
                                update(attackBonusMods, {
                                    [key]: { $set: convertInputValue(e.target.value) }
                                })
                            )
                        }}
                    />
                    <div>=</div>
                    <div>{attackBonusTotals[key].join(' / ')}</div>
                </Fragment>
            ))}

        </div>
    )
}


export default AttackBonuses