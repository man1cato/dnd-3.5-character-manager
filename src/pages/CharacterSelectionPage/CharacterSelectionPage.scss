@import '../../styles/base/settings';


.CharacterSelectionPage {
   margin-bottom: 0 !important;
}

.profile-button {
   display: flex;
   width: 100%;
   padding: $s-size;
   margin-bottom: $m-size;
   background: $off-white;
   border: 1px solid $light-grey;
   border-radius: $s-size;
   > img {
      width: 30%;
      max-height: 100px;
      margin-right: $s-size; 
   }
   :last-child {
      width: -webkit-fill-available;
      text-align: left;
      line-height: 2rem;
   }
}

button.profile-button:hover {
   box-shadow: 0 0 5px $light-grey;
}