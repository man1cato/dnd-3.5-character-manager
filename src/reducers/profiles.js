export default (state = [], action) => {
    switch (action.type) {
        case 'SET_PROFILES':
            return [
                ...action.profiles
            ]
        case 'DELETE_PROFILES':
            return []
        default:
            return state
    }
}
