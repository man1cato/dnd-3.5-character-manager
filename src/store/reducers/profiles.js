export default (state = [], action) => {
    switch (action.type) {
        case 'SET_PROFILES':
            return action.profiles
        case 'REMOVE_PROFILES':
            return []
        default:
            return state
    }
}
