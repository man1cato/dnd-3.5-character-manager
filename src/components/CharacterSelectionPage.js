import React from 'react'
import {connect} from 'react-redux';

import Header from './Header'
import {startSetProfile} from '../actions/profile'
import {history} from '../routers/AppRouter'


export class CharacterSelectionPage extends React.Component {
   state = {
      disabled: false 
   }

   handleClick = (e) => {
		const id = e.currentTarget.id
		this.setState(() => ({
			disabled: true
		}))
      this.props.startSetProfile(id)
      setTimeout(() => { history.push('/profile') }, 1500)
   }

   render() {
      return (
         <div className="layout">
            <Header pageTitle="Character Selection" />

            <div className="container container--body">
               <h3>Select your character:</h3>
               {this.props.profiles.map((profile, i) => (

                  <button 
                     key={i}
                     className="button-profile"
                     id={profile.id}   
                     disabled={this.state.disabled}
                     onClick={this.handleClick}
                  >
                     <img src={profile.iconUrl} />
                     <div>
                        <h3>{profile.name}</h3>
                        <div>{profile.gender} {profile.race}</div>
                        <div>{profile.jobClass}</div>
                        <div>{profile.alignment}</div>
                        <div>{`Level ${profile.level}`}</div>
                     </div>
                  </button>
               ))}

            </div>
         </div>
      )
   }
}

const mapStateToProps = (state) => ({
	profiles: state.profiles
})

const mapDispatchToProps = (dispatch, props) => ({
   startSetProfile: (id) => dispatch(startSetProfile(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(CharacterSelectionPage)