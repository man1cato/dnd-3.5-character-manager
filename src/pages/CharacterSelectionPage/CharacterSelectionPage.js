import React, { useState } from 'react'
import { connect } from 'react-redux'

import ConfirmationModal from '../../components/Modals/ConfirmationModal'
import { startSetProfile, startRemoveProfile } from '../../store/actions/profile'
import { history } from '../../routers/AppRouter'
import './CharacterSelectionPage.scss'


export const CharacterSelectionPage = props => {
   const [characterToRemove, setCharacterToRemove] = useState(null)
   const [disabled, setDisabled] = useState(false)

   const handleSelectCharacter = async profileId => {
      try {
         setDisabled(true)
         await props.startSetProfile(props.uid, profileId)
         history.push('/profile')
      } catch (e) {
         setDisabled(false)
         alert('An error occurred. Please try again.')
         console.log(e)
      }      
   }

   return (
      <div className="container--body CharacterSelectionPage">
         {props.profiles.map(profile => (
            <div className="profile-container" key={profile.id}>
               <button 
                  className="profile__select-button"
                  data-testid={profile.id}
                  disabled={disabled}
                  onClick={() => handleSelectCharacter(profile.id)}
               >
                  <img src={profile.iconUrl} />
                  <div>
                     <h3>{profile.name}</h3>
                     <div>{profile.gender} {props.races[profile.race].name}</div>
                     <div>{props.jobClasses[profile.jobClass].name}</div>
                     <div>{profile.alignment}</div>
                     <div>{`Level ${profile.level}`}</div>
                  </div>
               </button>
               <button
                  className="profile__remove-button"
                  data-testid={profile.id + 'RemoveButton'}
                  disabled={disabled}
                  onClick={() => setCharacterToRemove(profile)}
               >
                  <ion-icon name="trash" size="large"/>
               </button>
            </div>
         ))}
         <ConfirmationModal
            clickedItem={characterToRemove}
            messageTitle={`Delete Character ${characterToRemove && characterToRemove.name}?`}
            message="This action cannot be undone!"
            handleConfirm={() => props.startRemoveProfile(characterToRemove.id)}
            handleCloseModal={() => setCharacterToRemove(null)}
         />
      </div>
   )
}

const mapStateToProps = state => ({
   uid: state.auth.uid,
   profiles: state.profiles,
   jobClasses: state.jobClasses,
   races: state.races
})

const mapDispatchToProps = dispatch => ({
   startSetProfile: (uid, profileId) => dispatch(startSetProfile(uid, profileId)),
   startRemoveProfile: profileId => dispatch(startRemoveProfile(profileId))
})

export default connect(mapStateToProps, mapDispatchToProps)(CharacterSelectionPage)