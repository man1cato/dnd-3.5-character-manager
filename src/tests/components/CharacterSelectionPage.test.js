import React from 'react';
import { shallow } from 'enzyme';

import { CharacterSelectionPage } from '../../components/CharacterSelectionPage';
import profiles from '../fixtures/profiles'


const startSetProfile = jest.fn()
let wrapper

beforeEach(() => {
	wrapper = shallow(
		<CharacterSelectionPage 
			profiles={profiles} 
			startSetProfile={startSetProfile}
		/>
	)
})


test('should render CharacterSelectionPage with profiles data', () => {
	expect(wrapper).toMatchSnapshot()
})

test('should disable buttons and call startSetProfile', () => {
	const id = profiles[0].id 
	wrapper.find(`#${id}`).simulate('click', {
		currentTarget: { id }
	})
    
	expect(wrapper.state('disabled')).toBe(true)
	expect(wrapper.find('button').at(1).props().disabled).toBe(true)
	expect(startSetProfile).toHaveBeenCalledWith(id)
})