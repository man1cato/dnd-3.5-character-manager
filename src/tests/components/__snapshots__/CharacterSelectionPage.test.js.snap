// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharacterSelectionPage with profiles data 1`] = `
<div
  className="layout"
>
  <Connect(Header)
    pageTitle="Character Selection"
  />
  <div
    className="container container--body"
  >
    <h3>
      Select your character:
    </h3>
    <button
      className="button-profile"
      disabled={false}
      id="recnPKEA6sOWaOt57"
      key="0"
      onClick={[Function]}
    >
      <img
        src="images/elf.svg"
      />
      <div>
        <h3>
          Istvyll Bannianea
        </h3>
        <div>
          Male
           
          Grey Elf
        </div>
        <div>
          Wizard
        </div>
        <div>
          Lawful Evil
        </div>
        <div>
          Level 5
        </div>
      </div>
    </button>
    <button
      className="button-profile"
      disabled={false}
      id="-LfeNscfMYgOyQV11vXY"
      key="1"
      onClick={[Function]}
    >
      <img
        src="images/orc.svg"
      />
      <div>
        <h3>
          Korg
        </h3>
        <div>
          Male
           
          Half-Orc
        </div>
        <div>
          Wizard
        </div>
        <div>
          Lawful Evil
        </div>
        <div>
          Level 5
        </div>
      </div>
    </button>
  </div>
</div>
`;
